// @usecannon/cli/dist/src/chains.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainById = exports.chains = exports.cannonChain = void 0;
const viem = __importStar(require("viem"));
const viemChains = __importStar(require("viem/chains"));
exports.cannonChain = {
    id: 13370,
    name: 'Cannon Local',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18,
    },
    rpcUrls: { default: { http: ['http://localhost:8545'] } },
};
exports.chains = [exports.cannonChain, ...Object.values(viemChains)];
function getChainById(id) {
    const chain = viem.extractChain({
        chains: exports.chains,
        id,
    });
    if (chain)
        return chain;
    return {
        id,
        name: 'Unknown Network',
        nativeCurrency: {
            name: 'Ether',
            symbol: 'ETH',
            decimals: 18,
        },
        rpcUrls: { default: { http: [] } },
    };
}
exports.getChainById = getChainById;
