// @usecannon/builder/dist/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCannonRepoRegistryUrl = exports.DEFAULT_REGISTRY_CONFIG = exports.DEFAULT_REGISTRY_ADDRESS = exports.BUILD_VERSION = exports.CANNON_CHAIN_ID = void 0;
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('cannon:builder:constants');
exports.CANNON_CHAIN_ID = 13370;
exports.BUILD_VERSION = 7;
// Production settings (OP Mainnet & ETH Mainnet)
exports.DEFAULT_REGISTRY_ADDRESS = '0x8E5C7EFC9636A6A0408A46BB7F617094B81e5dba';
exports.DEFAULT_REGISTRY_CONFIG = [
    {
        name: 'OP Mainnet',
        chainId: 10,
        providerUrl: ['https://optimism-rpc.publicnode.com'],
        address: exports.DEFAULT_REGISTRY_ADDRESS,
    },
    {
        name: 'Ethereum Mainnet',
        chainId: 1,
        providerUrl: ['https://ethereum-rpc.publicnode.com'],
        address: exports.DEFAULT_REGISTRY_ADDRESS,
    },
];
/*
// Development settings (OP Sepolia & ETH Sepolia)
export const DEFAULT_REGISTRY_ADDRESS: Address = '0xEd58414AFA6706d2E19e009C635cC75D25A91f18';
export const DEFAULT_REGISTRY_CONFIG = [
  {
    name: 'OP Sepolia',
    chainId: 11155420,
    providerUrl: ['https://optimism-sepolia-rpc.publicnode.com'],
    address: DEFAULT_REGISTRY_ADDRESS,
  },
  {
    name: 'Ethereum Sepolia',
    chainId: 11155111,
    providerUrl: ['https://ethereum-sepolia-rpc.publicnode.com'],
    address: DEFAULT_REGISTRY_ADDRESS,
  },
];
*/
function getCannonRepoRegistryUrl() {
    const tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
    // to avoid DST issues we choose a time in the middle of the winter (obviousy this doesn't work for southern tzs but those arent relevant anyway)
    const shHistoricalOffset = new Date('2023-01-01').getTimezoneOffset();
    debug(`timezone for repo resolve is ${tz}`);
    let region = 'us-east';
    if (tz?.startsWith('Asia/')) {
        region = 'sg';
    }
    else if (tz?.startsWith('Australia/')) {
        region = 'au-east';
    }
    else if (shHistoricalOffset >= 360) {
        // there are a lot of america timezones, so we just check if its mountain time or later
        region = 'us-west';
    }
    debug('cannon repo region is ', region);
    return `https+ipfs://${region}.repo.usecannon.com`;
}
exports.getCannonRepoRegistryUrl = getCannonRepoRegistryUrl;
